evap/student/templates/student_vote.html

```diff
@@ -312,6 +312,259 @@ <h3 class="mb-3">{{ evaluation.full_name }} ({{ evaluation.course.semester.name
+                    // Don't do anything special
+                    return;
+                }
+
+                // Since the event could be caught on either the outer label or
+                // the nested text / input, the full row could be two steps up
+                let row = current.parentElement;
+                if (row.tagName !== "DIV") row = row.parentElement;
+                if (row.tagName !== "DIV") {
+                    return;
+                }
+
+                const insideSubmitRow = row.parentElement.classList.contains("card-submit-area");
+
+                if (!insideSubmitRow && current.tagName !== "TEXTAREA") {
+                    const num = parseInt(e.key);
+                    if (!isNaN(num)) {
+                        // Even if the number does not trigger a selection (for example pressing "9"),
+                        // nothing else should happen, because it would be very frustrating if only some numbers
+                        // would work as expected.
+                        e.preventDefault();
+                        selectByNumberKey(row, num);
+                        return;
+                    }
+                }
+
+                if (e.key !== "Tab" || e.shiftKey && focussable[0] === current) {
+                    // Go back to normal tabbing if you are on the top element and want to tab futher back.
+                    return;
+                }
+
+                e.preventDefault();
+                let nextElement;
+                const direction = e.shiftKey ? -1 : 1;
+
+                if (insideSubmitRow) {
+                    // Here, you can tab inside the row
+                    const nextIndex = Math.min(focussable.indexOf(current) + direction, focussable.length);
+                    nextElement = focussable[nextIndex];
+                } else {
+                    // Otherwise, there is work more to do.
+
+                    // The index of the nearest focusable element that is not in the current row
+                    let nextIndex = focussable.indexOf(current);
+                    if (nextIndex === -1) {
+                        nextElement = focussable[0];
+                    } else {
+                        while (0 <= nextIndex + direction
+                            && nextIndex + direction < focussable.length
+                            && focussable[nextIndex].parentElement.parentElement === current.parentElement.parentElement
+                        ) nextIndex += direction;
+
+                        // TODO: possible optimization: if focussable[nextIndex] is a textarea, just focus it
+                        // Is it worth it?
+                        const nextRow = focussable[nextIndex].parentElement.parentElement;
+
+                        if (nextRow.id === "few-participants-warning") {
+                            nextElement = nextRow.children[1].children[0];
```

Comment:
Hmm.... I feel like this will break at some point (especially this line, relying on the specific DOM to be this way, but also the whole script in general).

@janno42 @karyon @pixunil We already have some points where we do stuff on the front end that is currently untested. I feel like this code here is another reason to (finally) add front end tests -- or, to put it differently: I wouldn't feel confident having this without any tests. What do you think?

Reply:
I agree, all of this is pretty error prone. Maybe I could make this specifically more robust by introducing more CSS classes and querying here? Or maybe I should use xpath or something more expressive?

Reply:
Yes, having a possibility to unit-test (pure js functions) or integration-test (when you perceive html and js as two different code units) would be very nice. I had the urge to have some in my previous PR with the configurable text answers (where I just missed a `/g` flag in my regex).
I’m a bit overwhelmed with the research here, as the ecosystem is rather huge and I’m unsure whether it is sensible to migrate to e.g. Typescript or not when we are touching our javascript strategy. (but I think this general discussion is better discussed during a call)

----

For this specific line, I would prefer a `nextRow.querySelector(…)` if possible, if it is unique enough, you can match on the tag, otherwise adding a class would be the way to go.